crates/ruff_server/src/edit/document.rs

```diff
@@ -0,0 +1,120 @@
+use lsp_types::{Position, TextDocumentContentChangeEvent};
+use ruff_source_file::LineIndex;
+
+use crate::PositionEncoding;
+
+use super::range::text_range;
+
+#[derive(Debug, Clone)]
+pub struct Document {
+    contents: String,
+    index: LineIndex,
+    version: i32,
+}
+
+impl Document {
+    pub fn new(contents: String, version: i32) -> Self {
+        let index = LineIndex::from_source_text(&contents);
+        Self {
+            contents,
+            index,
+            version,
+        }
+    }
+    // TODO(jane): I would personally be in favor of removing access to this method and only
+    // allowing document mutation via specialized methods.
+    pub(crate) fn modify(&mut self, func: impl FnOnce(&mut String, &mut i32)) {
+        self.modify_with_manual_index(|c, v, i| {
+            func(c, v);
+            *i = LineIndex::from_source_text(c);
+        });
+    }
+
+    // A private function for overriding how we update the line index by default.
+    fn modify_with_manual_index(
+        &mut self,
+        func: impl FnOnce(&mut String, &mut i32, &mut LineIndex),
+    ) {
+        let old_version = self.version;
+        func(&mut self.contents, &mut self.version, &mut self.index);
+        debug_assert!(self.version >= old_version);
+    }
+}
+
+/* Mutable API */
+impl Document {
```

Comment:
I prefer not to group methods but I don't others doing it. What I don't understand is why `modify` isn't part of the `Mutable` api.